sklearn/datasets/_arff_parser.py

```diff
@@ -206,7 +207,13 @@ def _io_to_generator(gzip_file):
         # liac-arff parser does not depend on NumPy and uses None to represent
         # missing values. To be consistent with the pandas parser, we replace
         # None with np.nan.
-        frame = pd.concat(dfs, ignore_index=True).fillna(value=np.nan)
+        frame = pd.concat(dfs, ignore_index=True)
+        pd_version = parse_version(pd.__version__).base_version
+        if parse_version(pd_version) < parse_version("2.2"):
+            frame = frame.fillna(value=np.nan)
+        else:
+            with pd.option_context("future.no_silent_downcasting", True):
+                frame = frame.fillna(value=np.nan).infer_objects(copy=False)
```

Comment:
can we have this as a function in fixes.py? Rather keep all code depending on versions in the same place, makes it easier to maintain.

Reply:
I foresee that it will be super ugly because this linked to an optional dependency. But let see.

Reply:
Actually, I can pass `pd` so completely fine.